### ISSUE TITLE
mac + tmux + vim soooo slow 

### ISSUE THREAD

**Issue by dragonxlwang** [tmux/tmux#353]
I am using a mac. When using vim in tmux (actually macvim in iterm2/terminal),  vim motion becomes extra slow, no mention with mouth scrolling.

It seems to have something to do with vertical split windows in vim. If more than one window vertically split, then the slow is significant.

![untitled](https://cloud.githubusercontent.com/assets/15812510/14097706/a5295c28-f539-11e5-943a-8a25890fbe62.gif)

I have tried so far the following (1 being most slow while 10 being fastest):

1. Gui MacVin: 10
2. No tmux; iTerm2 / Terminal + Mac + MacVim: 8
3. No tmux; iTerm2 / Terminal + Mac + Apple's Vim: 6
4. Gnu-Screen + iTerm2 / Terminal + Mac + MacVim: 8
5. Tmux + iTerm2 / Terminal + Mac + MacVim (only 1 window): 4-5
6. Tmux + iTerm2 / Terminal + Mac + MacVim (only 2 window horizontally): 3-4
7. Tmux + iTerm2 / Terminal + Mac + MacVim (2 window vertically): 2
8. Tmux + iTerm2 / Terminal + Mac + MacVim (2+ window vertically): 1 (or even worse)

I admit that I have some plugins that I use a lot in vim. If I don't use any of them, then the speed becomes normal (10) again. If anyone want to reproduce, here's the list:

```
ListToggle
Vundle.vim
YCM-Generator
YankRing.vim
YouCompleteMe
auto-pairs
ctrlp.vim
gist-vim
indentLine
nerdtree
nerdtree-git-plugin
tabular
tagbar
ultisnips
vim-airline
vim-airline-themes
vim-autoformat
vim-bbye
vim-better-whitespace
vim-bracketed-paste
vim-coffee-script
vim-colorschemes
vim-commentary
vim-eunuch
vim-flake8
vim-fugitive
vim-gitgutter
vim-minimap
vim-snippets
vim-surround
vim-syntax-extra
webapi-vim
```

**Comment by ntnn** [comment]
Eliminate plugin by plugin until you find the culprit. Not really a tmux issue when the problem is with a program where you yourself said that it works fine without plugins.

**Comment by dragonxlwang** [comment]
I tried... even with simple plugin like vim-color-schemes the speed would slow down greatly.

**Comment by ntnn** [comment]
Either way, this is not a tmux issue. I'd suggest you go to vim bugtracker and ask there for support.
And even there they will tell you to eliminate one plugin after the other until you know what is going on (or the other way around - add plugin by plugin until you found something).

**Comment by dragonxlwang** [comment]
I hope I have made it clear --- no matter which one single plugin I tried enabled, it slows down  suddenly. I doubt this is related to any single plugins at all.

**Comment by ntnn** [comment]
Another helpful thing would be your configuration files.

**Comment by dragonxlwang** [comment]
Thanks, I will look into it… I even set escape time to be 0

https://gist.github.com/anonymous/5c50e58d9378f96d1f5a <https://gist.github.com/anonymous/5c50e58d9378f96d1f5a>

**Comment by ntnn** [comment]
> # use "xterm-256color" instead of "screen-256color"
> set -g default-terminal "xterm-256color"

You don't read documentation at all, do you?
Set TERM to an appropriate value, which is either a variant of `screen`
or `tmux`, depending on how recent your ncurses is.

**Comment by dragonxlwang** [comment]
sorry ...  I have no idea which documentation you are referring to. But you are right, if I change it to screen-256color it becomes better though not as fast as in gui macvim.

I remember I set it to xterm-256color to be compatible with iterm...

Could you give me a pointer?

**Comment by ntnn** [comment]
Manual:

	default-terminal terminal
		Set the default terminal for new windows created in this
		session - the default value of the TERM environment vari-
		able.  For tmux to work correctly, this must be set to
		`screen', `tmux' or a derivative of them.

FAQ (right at the start):

    tmux frequently asked questions

    ******************************************************************************
    * PLEASE NOTE: most display problems are due to incorrect TERM! Before       *
    * reporting problems make SURE that TERM settings are correct inside and     *
    * outside tmux.                                                              *
    *                                                                            *
    * Inside tmux TERM must be "screen" or similar (such as "screen-256color").  *
    * Don't bother reporting problems where it isn't!                            *
    *                                                                            *
    * Outside, it must match your terminal: particularly, use "rxvt" for rxvt    *
    * and derivatives.                                                           *
    ******************************************************************************

Your problem most likely entirely comes from heavy misconfiguration.

**Comment by dragonxlwang** [comment]
My bad --- never saw that faq before. I installed tmux from homebrew...  i will see what's wrong with xterm there..

**Comment by dragonxlwang** [comment]
@ntnn, sorry for my carelessness of not reading the FAQ first!

I went over again the manual which is really helpful. But I am still not very clear why xterm is not allowed to be set "inside" tmux as when search .tmux.conf in github there are soooo many people using xterm-256colors yet (which I assume they are wrong?)

**Comment by ntnn** [comment]
@dragonxlwang yes, they are doing it wrong. Screens' and xterms'
terminfo are close enough not to cause problems all the time, but they
differ greatly.

You'll also have to check what your term is outside of tmux. It has to
match what your term actually uses as terminfo (this should be set
correctly by default).

After that you'll have to go over your plugins again, disabling/enabling
one by one, and I'd need to see your vim configuration to see if
something is off there.

**Comment by dragonxlwang** [comment]
> You'll also have to check what your term is outside of tmux. It has to
> match what your term actually uses as terminfo (this should be set
> correctly by default).

Well, this is one part I am a bit confused. Not sure what exactly it means by "outside tmux" I saw that a lot in FAQ as well. I am using iTerm2, and if not using tmux, TERM is set to xterm-256colors. How can I match term outside tmux? should I alias tmux="TERM xterm-256colors tmux"? Or I am already all set?

**Comment by josefnpat** [comment]
_There's a lot of patience in this thread_

**Comment by dragonxlwang** [comment]
@josefnpat  sorry if you think I've abused this place ...

**Comment by ntnn** [comment]
> Well, this is one part I am a bit confused. Not sure what exactly it
> means by "outside tmux" I saw that a lot in FAQ as well. I am using
> iTerm2, and if not using tmux, TERM is set to xterm-256colors. How can
> I match term outside tmux? should I alias tmux="TERM xterm-256colors
> tmux"? Or I am already all set?

'Outside of tmux' means the environment you're running tmux in, so in
your case iTerm2 and the shell you're using.

That the term env needs to match your emulator means that you don't
fiddle with the term env - the emulator should set it correctly for you
already (and if it doesn't its time to use another one).

Note:
You'll also see people exporting things, such as term, in `.bashrc` or
`.zshrc` or similar - that will break things in the long run.
Exports solely belong in the login shell init file, such as
`.bash_profile` or `.zprofile` and even then you don't export some
things like term.

> Yes. Hope this can be solved by remove only a few plugins.

I suspect nerdtree, airline, youcompleteme and anything that has to do
with syntax. Try those first.

**Comment by dragonxlwang** [comment]
@ntnn, this is the most clear explanation I have ever seen about term settings. I was always confused why half people suggest export term in .xxxrc while others say iterm/terminal should do that for me and I should leave it as is. I even found on stackoverflow people say it should set xterm-256colors in tmux.conf. Guess I should never take words for granted without looking manually in details... Thanks again! Now by only setting term screen-256colors in tmux.conf while not touching $TERM anywhere else it seem working fine. I will plugin/plugout vim/tmux packages one by one from now...

**Comment by dragonxlwang** [comment]
So far, I have done an extensive experiment with tmux+vim. For the following minimal configuration I still see the lag when scrolling more than 200-300 lines.

In my .tmux.conf, I have 
```sh
set -g default-terminal "screen-256color" 
```
and use any 256 colorscheme;

While I use vimrc from [here](https://github.com/amix/vimrc/blob/master/vimrcs/basic.vim) (without any plugin)

So, IMHO this is an issue related to 256 colors in tmux + vim. If not using 256 there is no problem anymore. But if use 256colors, it becomes slower when using more display such as:

```viml
set number                      " Line numbers are good
set ruler                           " Bottom right corner of the status line
set textwidth=80              " Maximum line text width
set colorcolumn=80         " Highlight column at 80 char
set visualbell                    " No sounds
```

I understand tmux is not supposed to have vim behave as fast as gui vim, but it seems those problems related to 256 colors are not appeared in screen. It would be great to know how to improve this. Thanks,

**Comment by nicm** [comment]
is this the same in xterm?

**Comment by vaxXxa** [comment]
+1 for this issue.

vim outside tmux works fine.. in tmux i have lag during scroll big files (~1000 LOC)

**Comment by ntnn** [comment]
@vaxXxa Information on your particular system and configuration would be helpful. Especially TERM inside and outside of tmux, pastebins of your tmux and vim configuration, verified not working terminals, with and without syntax highlightning, with `vim -uNONE $file`, ...

Also, profile your vim configuration (`:h profiling`) and see if there's a notable difference in the conclusion at the bottom inside and outside of tmux (please pastebin these too).

**Comment by tweekmonster** [comment]
FWIW, tmux performs very poorly in iTerm after `3.0.20160713 nightly`: https://gitlab.com/gnachman/iterm2/issues/4976

**Comment by scarfacedeb** [comment]
@dragonxlwang I was also having performance and redraw issues with a similar configuration.
Removing `set -g focus-events on` from tmux.conf resolved them for me. (It was set by tmux-sensible plugin).

**Comment by Kutsan** [comment]
Same. :/

**Comment by itsNikolay** [comment]
The same stuff with iTerm2 3.013 + tmux 2.3 + vim 8.0
Anybody eliminated the issue ?

**Comment by eleijonmarck** [comment]
It seems as the key "ESC" is espacially slow. The other key strokes seems to be flawless. Is this tied to the special character flow to tmux?

**Comment by dragonxlwang** [comment]
I finally gave this up in digging the reasons. It seems that a lot of different plugins will slow tmux+vim down. There are three "solutions" probably would work:

1. disable vim plugins
2. don't use tmux 
3. find a better machine 

I chose 3 and the slow-down has decreased to a bearable extent.  :)

**Comment by krader1961** [comment]
This is due to an unfortunate choice by the original vi creator to use [esc] as the character for transitioning from insert to normal mode. Since most of the non-alphanumeric keys send a sequence of characters that begin with an escape it is necessary for programs like tmux and bash (or whatever shell you use) to wait briefly after seeing an escape to see if more characters that represent a special key like [left-arrow] arrive. See the `escape-time` variable. Most people who use vi/vim inside tmux have something like the following in our *~/.tmux.conf*:

```
set-option -s escape-time 10
```

Some people set it to zero but I consider that risky if you're connecting over a wide-area network or there is anything else that might insert small delays between the delivery of chars in such a sequence.

**Comment by eleijonmarck** [comment]
@krader1961 waiiitttt what??? That solved my issue. Thank you for the explanation. Now I understand why the it took longer for escape to work.

**Comment by krader1961** [comment]
FWIW, I addressed a similar issue with the [fish shell](https://github.com/fish-shell/fish-shell) roughly a year ago. Although in that case it necessitated increasing the default timeout after seeing an escape character. With respect to tmux the default of 500 ms should be decreased. But no matter what value the respective programs use there will always be problems. The problem is inherent in the way that vi/vim use the escape character (i.e., the [esc] key).

**Comment by rajaravivarma-r** [comment]
Reducing the **Scrollback buffer** helped me resolve the scroll and cursor movement lag. Also, it improved the general tab completion lag in Mac Terminal.

In my case I changed the scrollback setting from `Limit to available memory` to `Limit number of rows to [1000]`.

Source (not my blog): https://arvid.io/2016/01/26/fixing-slow-vim-in-tmux-and-iterm2/

Note: I have also set `set-option -s escape-time 10` suggested by @krader1961.

**Comment by mjacobus** [comment]
Helped for me:

set norelativenumber
set nocursorline

**Comment by jcconnell** [comment]
I'm only using Vim on iTerm2 with some light modifications. No Tmux. @mjacobus your suggestion helped tho, thanks!

**Comment by thenewvu** [comment]
Hey guys, I tried below workarounds and now I don't feel slowing when scrolling anymore:

```
" disable auto matching parens
let g:loaded_matchparen=1
" don't render special chars (tabs, trails, ...)
set nolist
" if number column doesn't make sense to you, just don't render it
set nonumber
" this is common things
set lazyredraw
set ttyfast
```

And a weird thing is, try to change to another font can optimize rendering, I've used `Operator Mono`, rendering it much slower than other fonts, let try `SF Mono 14`, you will see.

Hope this will be useful for you guys.

**Comment by timvisee** [comment]
@thenewvu Setting the font to something different without anti-aliasing on macOS helps quite a lot. Everything is suddenly fast now.

I have to say, what an horrible OS that things like fancy font rendering are so slow, it's slowing down everything, when using a full-screen terminal application.

**Comment by jcconnell** [comment]
@timvisee I experimented with this setting a bit this morning. I had recently changed my font to a nerd font and noticed that the terminal was slowing down again. Switching off anti-aliasing sped it up again. 

I use a large 4k display though and without anti-aliasing the fonts look terrible.

**Comment by nyjt** [comment]
I had the same issue and I've started using [alacritty](/jwilm/alacritty) terminal. Now vim is insanely fast on my 32" 4k display.

**Comment by timvisee** [comment]
@nyjt Alacritty is awesome ineed. Although I'm still experiencing problems on macOS with keys such as `~` with the language/keyboard settings I'm using. I've already switched to it on my Linux machine, but I might have to wait a little more on macOS.

**Comment by brianarpie** [comment]
I was experiencing seriously lag using vim inside of tmux. I tried profiling, removing slow plugins, applying settings suggested in this thread. Some of these suggestions helped but I could not completely get rid of the lag when running vim inside tmux.

Finally, I decided to boot up iTerm2. Vim + Tmux + Mac is now lightning fast. I switched to Terminal a few months ago being told it was more performant. That was not true in my case.

I hope this helps someone else who may be experiencing a slow Vim + Tmux + Mac experience.

**OSX**: 10.12.6
**Tmux**: 2.6
**Vim**: 8.0.304
**iTerm2**: 3.1.5
**Terminal**: 2.7.3

**Comment by danemacmillan** [comment]
In my experience, `set cursorline` is the single biggest performance hog in vim. Take it out, or configure `set nocursorline` and scroll lag will disappear.

**Comment by pencilcheck** [comment]
Second @brianarpie somehow Terminal + tmux + neovim is super sluggish when scrolling. It literally feels like the CPU is bursting with tears when I am trying to move the cursor. I remove the cursorline, I remove plugin by plugin, still get slower as time goes on.

An immediately night and day difference once I quit tmux and open vim in Terminal.app directly.

Right now going back to iTerm2, the performance not the best but faster than Terminal so far. Will be continuing using it for an extended amount of time to see if it gets slower as well.

**Comment by joshdover** [comment]
If you have a Mac with a discrete graphics card and you're using iTerm 2, play around with the "Advanced GPU Settings". I had some success disabling the "Prefer integrated to discrete GPU" option.

![image](https://user-images.githubusercontent.com/1813008/44793819-c5506c80-ab6c-11e8-9702-5bc232128be0.png)

**Comment by bradlarsen** [comment]
tl;dr: try speeding up your keyboard repeat rate; bad vim redrawing performance may have nothing to do with tmux.

For posterity:

I had been suffering through slow mac + vim + tmux performance for a couple years on a Retina iMac, and finally got around to investigating and resolving my troubles!

I did a whole bunch of informal experiments involving scrolling through a ~500 line file by holding down 'j' in vim, and using a stopwatch to time how long it took to reach the bottom.  The factors I investigated:

- Terminal.app vs iTerm.app vs Alacritty
- tmux vs no tmux
- vim 8 vs neovim
- vim/neovim with no extra config files (`-u NONE`) vs using my plugin-heavy config
- Menlo vs Monaco
- Aliased vs Antialiased font
- Scaled vs non-scaled screen resolution
- `lazyredraw` and `ttyfast` in vim/neovim
- wired vs wireless keyboard
- keyboard repeat rate

For me, the only factor that made any noticeable difference in my scroll-to-the-bottom experiment was adjusting the keyboard repeat rate!  I had already had the initial delay and keyboard repeat rates set to their minimum through System Preferences, and then found [this helpful Stack Exchange post about how to speed up these settings even further](https://apple.stackexchange.com/questions/10467/how-to-increase-keyboard-key-repeat-rate-on-os-x).  After I set KeyRepeat to 1 (twice as fast as the fastest rate exposed in System Preferences), my scroll-to-the-bottom experiment became twice as fast, and the rendering is soooo much smoother.  With or without my plugin-heavy vim configuration, it seems like only the keyboard repeat rate affects how smoothly vim scrolls and renders for me.  Go figure.

**Comment by jcconnell** [comment]
@bradlarsen Thanks! Your post intrigued me so I changed KeyRepeat right away. I was already using a suggestion from earlier in this thread, but your improvement definitely makes it feel more snappy. I'll report back after some more use.

**Comment by slack0** [comment]
> @krader1961 waiiitttt what??? That solved my issue. Thank you for the explanation. Now I understand why the it took longer for escape to work.
> 
> @dragonxlwang this solved my issues. Maybe this might be some what people have been looking for.

I have suffered long enough with this exact same problem. Just adopted this solution -- and everything just got snappier. I was on the verge of switching away from Vim + Tmux (unthinkable!) -- to  avoid this annoying slow scrolling issue. 

The issue is not so much with the slow scroll; it is with the cognitive dissonance it induces into the thought process.

Glad I stumbled across this issue.
